src/main/java/com/github/shyiko/mysql/binlog/io/BufferedSocketInputStream.java

```diff
@@ -61,10 +61,10 @@ public int read(byte[] b, int off, int len) throws IOException {
             offset = 0;
             limit = in.read(buffer, 0, buffer.length);
         }
-        int bytesRemainingInBuffer = Math.min(len, limit - offset);
+        int bytesRemainingInBuffer = Math.min(len, Math.max(limit - offset, 0));
```

Comment:
How about to return early if `bytesRemainingInBuffer` is 0? No need to copy then or update the offset, right?